awscli/autocomplete/autogen.py

```diff
@@ -0,0 +1,177 @@
+"""This module is used to auto-generate server side completion data.
+
+This module can be used during the build process to generate any new
+server side autocompletions based on updated service models.
+It can also be used to regen completion data as new heuristics are added.
+
+"""
+import logging
+from difflib import SequenceMatcher
+from collections import namedtuple
+
+
+LOG = logging.getLogger(__name__)
+Resource = namedtuple('Resource', ['resource_name', 'ident_name',
+                                   'operation', 'jp_expr'])
+
+
+class ServerCompletionHeuristic(object):
+    # Prefix that typically indicates listing resources,
+    # e.g ListResources, DescribeResources, etc.
+    _RESOURCE_VERB_PREFIX = ('list', 'describe')
+
+    def __init__(self, singularize=None):
+        if singularize is None:
+            singularize = BasicSingularize()
+        self._singularize = singularize
+
+    def generate_completion_descriptions(self, service_model):
+        """
+
+        :param service_model: A botocore.model.ServiceModel.
+        """
+        # This method is a best-effort attempt at generating server side
+        # auto-completion data based on a set of heuristics.  It's not
+        # completely accurate and the results should be manually vetted.
+
+        # candidate -> a potential operation that can be used to
+        #   to describe a resource (e.g ListResources, DescribeResources)
+        # non_candidate -> a potential operation that can use
+        #   auto-completion data (e.g DeleteResource, UpdateResource)
+        candidates = []
+        non_candidates = []
+        for op_name in service_model.operation_names:
+            if op_name.lower().startswith(self._RESOURCE_VERB_PREFIX):
+                candidates.append(op_name)
+            else:
+                non_candidates.append(op_name)
+        all_resources = self._generate_resource_descriptions(
+            candidates, service_model)
+        all_operations = self._generate_operations(non_candidates,
+                                                   all_resources,
+                                                   service_model)
+        return {
+            'resources': all_resources,
+            'operations': all_operations,
+        }
+
+    def _generate_resource_descriptions(self, candidates, service_model):
+        all_resources = {}
+        for op_name in candidates:
+            resource = self._resource_for_single_operation(
+                op_name, service_model)
+            if resource is not None:
+                self._inject_resource(all_resources, resource)
+        return all_resources
+
+    def _generate_operations(self, op_names, resources, service_model):
+        # member_name -> resource
+        reverse_mapping = {}
+        for name, identifier_map in resources.items():
```

Comment:
It would be nice to break this up into smaller internal methods for readability as well